Source/DafnyLanguageServer/Workspace/ProjectManagerDatabase.cs

```diff
@@ -135,43 +146,49 @@ public class ProjectManagerDatabase : IProjectDatabase {
       lock (myLock) {
         var projectManagerForFile = managersBySourceFile.GetValueOrDefault(uri);
 
+        if (projectManagerForFile is { IsDisposed: true }) {
+          // Defensive coding
+          logger.LogError("Found disposed project manager through managersBySourceFile.");
+          projectManagerForFile = null;
+        }
+
+        bool triggerCompilation = false;
         if (projectManagerForFile != null) {
+
           var filesProjectHasChanged = !projectManagerForFile.Project.Equals(project);
           if (filesProjectHasChanged) {
+            logger.LogDebug($"Migrating file {uri} from project {projectManagerForFile.Project.Uri} to project {project.Uri}");
+
             var projectFileContentHasChanged = projectManagerForFile.Project.Uri == project.Uri;
             if (projectFileContentHasChanged) {
-              // Scrap the project manager.
-              projectManagerForFile.CloseAsync();
               managersByProject.Remove(project.Uri);
-            } else {
-              var previousProjectHasNoDocuments = projectManagerForFile.CloseDocument(projectManagerForFile.Project.Uri);
-              if (previousProjectHasNoDocuments) {
-                // Enable garbage collection
-                managersByProject.Remove(projectManagerForFile.Project.Uri);
-              }
-              logger.LogDebug($"Migrating file {uri} to existing project {project.Uri}");
+            }
+            var previousProjectHasNoDocuments = projectManagerForFile.CloseDocument(projectManagerForFile.Project.Uri);
+            if (previousProjectHasNoDocuments) {
+              // Enable garbage collection
+              managersByProject.Remove(projectManagerForFile.Project.Uri);
             }
 
             projectManagerForFile = managersByProject.GetValueOrDefault(project.Uri) ??
                                     createProjectManager(scheduler, verificationCache, project);
             projectManagerForFile.OpenDocument(uri, true);
+            triggerCompilation = true;
           }
         } else {
           var managerForProject = managersByProject.GetValueOrDefault(project.Uri);
+          if (managerForProject is { IsDisposed: true }) {
+            // Defensive coding
+            logger.LogError("Found disposed project manager through managersByProject.");
+            managerForProject = null;
+          }
+
           if (managerForProject != null) {
             projectManagerForFile = managerForProject;
-            if (changedOnOpen) {
-              projectManagerForFile.UpdateDocument(new DidChangeTextDocumentParams {
```

Comment:
Where did this code go? Why is it safe to remove?

Reply:
Replaced by the `projectManagerForFile.OpenDocument(uri, triggerCompilation);` call at the bottom of the method.